vm/systemSmartContracts/auction.go

```diff
@@ -0,0 +1,748 @@
+	return nil
+}
+
+func (s *stakingAuctionSC) unStake(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	registrationData, err := s.getRegistrationData(args.CallerAddr)
+	if err != nil {
+		return vmcommon.UserError
+	}
+
+	blsKeys, err := getBLSPublicKeys(registrationData, args)
+	if err != nil {
+		return vmcommon.UserError
+	}
+
+	for _, blsKey := range blsKeys {
+		stakedData, err := s.getStakedData(blsKey)
+		if err != nil || len(stakedData.RewardAddress) == 0 {
+			log.Debug("bls key was not staked")
+			return vmcommon.UserError
+		}
+
+		if !stakedData.Staked {
+			log.Debug("bls key was already unstaked")
+			return vmcommon.UserError
+		}
+
+		stakedData.Staked = false
+		stakedData.UnStakedNonce = s.eei.BlockChainHook().CurrentNonce()
+		err = s.saveStakedData(blsKey, stakedData)
+		if err != nil {
+			log.Debug("error while saving staked data")
+			return vmcommon.UserError
+		}
+	}
+
+	return vmcommon.Ok
+}
+
+func getBLSPublicKeys(registrationData *AuctionData, args *vmcommon.ContractCallInput) ([][]byte, error) {
+	blsKeys := registrationData.BlsPubKeys
+	if len(args.Arguments) > 0 {
+		for _, argKey := range args.Arguments {
+			found := false
+			for _, blsKey := range blsKeys {
+				if bytes.Equal(argKey, blsKey) {
+					found = true
+					break
+				}
+			}
+
+			if !found {
+				log.Debug("bls key for validator not found")
+				return nil, vm.ErrBLSPublicKeyMissmatch
+			}
+		}
+
+		blsKeys = args.Arguments
+	}
+
+	return blsKeys, nil
+}
+
+func (s *stakingAuctionSC) deleteBLSStakedData(blsKey []byte) (*big.Int, error) {
+	stakedData, err := s.getStakedData(blsKey)
+	if err != nil || len(stakedData.RewardAddress) == 0 {
+		log.Debug("bls key was not staked")
+		return nil, vm.ErrBLSKeyIsNotStaked
+	}
+
+	if stakedData.Staked || stakedData.UnStakedNonce < stakedData.StartNonce {
+		log.Debug("unBound is not possible for address which is staked or is not in unbound period")
+		return nil, vm.ErrStillInUnBoundPeriod
+	}
+
+	currentNonce := s.eei.BlockChainHook().CurrentNonce()
+	if currentNonce-stakedData.UnStakedNonce < s.unBoundPeriod {
+		log.Debug("unBound is not possible for address because unbound period did not pass")
+		return nil, vm.ErrStillInUnBoundPeriod
+	}
+
+	s.eei.SetStorage(blsKey, nil)
+	config := s.getConfig(stakedData.UnStakedEpoch)
+
+	return config.NodePrice, nil
+}
+
+func (s *stakingAuctionSC) unBound(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	registrationData, err := s.getRegistrationData(args.CallerAddr)
+	if err != nil {
+		return vmcommon.UserError
+	}
+
+	blsKeys, err := getBLSPublicKeys(registrationData, args)
+	if err != nil {
+		return vmcommon.UserError
+	}
+
+	totalUnBound := big.NewInt(0)
+	for _, blsKey := range blsKeys {
+		nodePrice, err := s.deleteBLSStakedData(blsKey)
+		if err != nil {
+			return vmcommon.UserError
+		}
+
+		ownerAddress := s.eei.GetStorage([]byte(ownerKey))
+		err = s.eei.Transfer(args.CallerAddr, ownerAddress, nodePrice, nil)
```

Comment:
This order should be changed. Otherwise the caller ends up sending wallet funds toward the owner of this SC.

Reply:
actually it was correct as the first is the destination. :P